### PR TITLE
Scope the tables selector per connection, not per data source

### PR DESCRIPTION
**Pull request by yochze**
An agent with 2+ connections often showed only one connection's tables.

`get_data_source_schema_paginated` decided ONCE per data source, from
`data_source.connections[0]`, whether to scope rows to the caller's
per-user overlay. Two failures fell out of that, and which one you got
depended on the order the database happened to return the connections in
— the relationship had no ORDER BY, so it could differ between requests
on the same agent:

  * delegated connection first — the whole catalog was filtered through
    an overlay that only ever described that one connection (the sync
    built its client with `construct_client`, documented as "first
    connection only"), so every other connection's tables vanished from
    the rows, the counts and the filter dropdowns. This is the reported
    bug.

  * delegated connection second — no scoping was applied at all, so one
    user saw another user's delegated Power BI models.

Resolve the scope per connection instead (`_resolve_catalog_scope`):
each connection lands in open / overlay-scoped / denied on its own auth
policy and this caller's own access, and every count, page, total and
filter dropdown runs through the same predicate, so a row can no longer
be hidden from the grid while still being counted in "Showing 1-20 of N"
or named in the schema dropdown.

The overlay itself is now connection-aware end to end:

- `user_data_source_tables` gains `connection_id` (backfilled through
  connection_tables; unique key widened to include it, so two
  connections exposing the same table name no longer collide).
- `get_user_data_source_schema` fans out over every per-user connection
  instead of building one client for `connections[0]`, and
  `_upsert_user_overlay` reconciles per connection — syncing one
  connection no longer revokes another's rows.
- Canonical-row lookup is connection-scoped, so an `orders` on one
  connection can't link to an `orders` on another.
- Rows a delegated user discovered themselves (no ConnectionTable to
  link to) carry their connection as provenance, so they render with the
  right connection and the connection filter reaches them.

The legacy non-paginated `get_data_source_schema` returned EITHER the
overlay OR the canonical catalog; on a mixed agent that dropped the
other connections. It now merges them — overlay tables (which carry
per-user column masking) for delegated connections, canonical tables for
the open ones.

Also: per-connection live connection tests in
`build_user_status_for_connection` (it was reporting connection A's
reachability as connection B's status), and a deterministic `order_by`
on `DataSource.connections` so the remaining `connections[0]` callers are
at least reproducible.

Verified end to end against a real Postgres, a real SQLite and a real
delegated Power BI/Entra connection on one agent, with two signed-in
users: before, a member saw 18 tables (Power BI only); after, 27 (5
Postgres + 4 SQLite + 18 Power BI), with each user seeing only their own
Power BI models.

Co-Authored-By: Claude Opus 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01Ks6uuXjodLpo8TFCFfQMJm

### CODE REVIEW

**Comment by yochze** [comment]
Reviewed commit `68833f3d1c0ec0f02c42271e841b0484024ab0b4`. I recommend changes before merging: six findings, including two access-control regressions.

1. **[P1] Full-schema merge loses delegated column masking.** `read_user_data_source_schema()` constructs `Table` objects without IDs (`id` defaults to `None`). The new filter discards every overlay, then substitutes canonical tables with the broader column set. Reproduced through `get_data_source_schema`: an external catalog exposing only `id` returned both `id` and `restricted_salary` from the canonical schema. Populate canonical IDs on overlay DTOs and restrict canonical fallback to open connections, including when overlay reading fails. [Lines 3147–3165](https://github.com/bagofwords1/bagofwords/blob/68833f3d1c0ec0f02c42271e841b0484024ab0b4/backend/app/services/data_source_service.py#L3147-L3165)

2. **[P1] Denied connections still expose previously discovered models.** When `overlay_ids` is empty, `overlay_rows` has no connection restriction; the unlinked-row branch still admits those stale accessible overlays even when their connection is denied. Reproduced through the selector service: after deactivating the member's credentials, `Private/T9` remained visible alongside the warehouse tables. Require current connection authorization for unlinked rows too. [Lines 3420–3454](https://github.com/bagofwords1/bagofwords/blob/68833f3d1c0ec0f02c42271e841b0484024ab0b4/backend/app/services/data_source_service.py#L3420-L3454)

3. **[P2] Dataset identity matching still crosses connections.** The new connection filter scopes `canonical_by_name`, but `canonical_by_dataset_table` still indexes every canonical row and takes precedence over the name lookup. Reproduced by syncing the same Power BI dataset/table through two delegated connections: both connection-specific overlays pointed to the same canonical row. This loses independent connection attribution, activation, and filtering. Apply connection ownership to both lookup indexes. [Lines 4559–4587](https://github.com/bagofwords1/bagofwords/blob/68833f3d1c0ec0f02c42271e841b0484024ab0b4/backend/app/services/data_source_service.py#L4559-L4587)

4. **[P2] First-read warming skips newly attached delegated connections.** Any existing overlay short-circuits warming for the whole agent. Reproduced: warm A, attach B with valid user credentials, then read the schema; B still has no overlay. Its catalog remains absent until an explicit refresh or another sync trigger. Check for missing overlays per connection. [Lines 3181–3188](https://github.com/bagofwords1/bagofwords/blob/68833f3d1c0ec0f02c42271e841b0484024ab0b4/backend/app/services/data_source_service.py#L3181-L3188)

5. **[P2] The roster omits IDs required by its advertised discovery call.** The roster emits names and numeric aliases, but `DescribeTablesInput.connection_ids` and the schema builder require UUIDs; there is no alias/name resolver. Reproduced with three connections and a one-table sample: all names appeared, but only the sampled connection's UUID was present in the complete rendered context. The agent cannot reliably scope discovery to the other connections, especially with duplicate table names. Include stable UUIDs in the roster and its tail. [Lines 1023–1039](https://github.com/bagofwords1/bagofwords/blob/68833f3d1c0ec0f02c42271e841b0484024ab0b4/backend/app/ai/context/sections/tables_schema_section.py#L1023-L1039)

6. **[P2] Downgrade fails on valid cross-connection duplicates.** Restoring `uq_user_ds_table` fails once two connections expose the same table name for the same user and agent. Reproduced with the actual migration: upgrade/backfill and repeat upgrade succeed; insert the now-valid duplicate on another connection; downgrade raises `UNIQUE constraint failed` while restoring the old constraint. Define a safe reconciliation policy before recreating that constraint. [Lines 118–121](https://github.com/bagofwords1/bagofwords/blob/68833f3d1c0ec0f02c42271e841b0484024ab0b4/backend/alembic/versions/overlayconn01_per_connection_user_overlay.py#L118-L121)

Validation: the five focused test files passed **28 tests**. Eight adjacent files produced **38 passed, 2 failed**. Both failures are in `test_refresh_overlay_prefetch.py` (`test_prefetched_tables_skip_live_fetch`, `test_no_prefetch_falls_back_to_live_fetch`), which still expect the old single-client path. All **3 tests in that file pass at the PR merge-base**, `4717340cf8876e00d69b234df3079c843d7044d4`, confirming introduced test regressions. Update the fixtures/stubs while preserving the prefetch behavior contract.

Five additional isolated checks reproduced findings 1–5 using the real services/database and a fake external schema client (the roster check exercises rendering). Finding 6 used a self-contained migration harness. Validation was on SQLite; PostgreSQL, live-provider behavior, and browser QA were not run.

File: backend/app/services/data_source_service.py [hunk 1]
```diff
@@ -3088,66 +3105,101 @@ async def get_data_source_schema(self, db: AsyncSession, data_source_id: str, in
...
+            return []
+
+        overlay_tables = []
+        try:
+            overlay_tables = [
+                t for t in await self.read_user_data_source_schema(
+                    db=db, data_source=data_source, user=current_user,
+                    active_only=not include_inactive,
+                )
+                if str(getattr(t, "id", "")) in visible_ids
```
[the remark this answers is not in the document]
Comment (yochze) [reply backend/app/services/data_source_service.py]: Confirmed and fixed in b54a480. Reproduced first: as a delegated member, `read_user_data_source_schema` returned 19 tables, 0 of which carried an id, so the filter kept 0 and all 19 Power BI tables came back from the canonical branch.

Two changes:

- The DTO now carries the canonical id (`data_source_table_id`), plus the connection id.
- Canonical is no longer a general fallback. It serves only rows attributable to an **open** connection, plus unlinked legacy rows no delegated user's own sync contributed — so a failed overlay read costs the delegated tables instead of unmasking them.

Covered by `tests/e2e/test_overlay_review_regressions.py::test_full_schema_serves_the_users_masked_columns` and `::test_overlay_read_failure_does_not_fall_back_to_canonical`, both of which fail at 68833f3.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yochze** [comment]
Two checks are red on `b54a480` and neither is this PR's. Both have the same root cause: **the OpenAI account has no credits.**

**`integration-llms`** — 9 failed, 6 passed, 14 skipped. Every failure is the same error, all on the `[openai]` params (`[anthropic]` and the rest pass):

```
openai.RateLimitError: Error code: 429 - {'error': {'message': 'You have no credits
remaining. Add credits to continue using the API at
https://platform.openai.com/settings/organization/billing/.',
'type': 'insufficient_quota', 'param': None, 'code': 'credit_balance_exhausted'}}
```

**`playwright-tests`** — 1 failed, 3 passed, 84 did not run. The failure is `onboarding-wizard.spec.ts:94`, waiting for `/successfully connected/i` on **step 2: configure OpenAI LLM provider**. That step runs a live connection test against OpenAI, so it is the same exhausted account surfacing through the UI.

Both jobs also fail on `main` — run [34688925450](https://github.com/bagofwords1/bagofwords/actions/runs/34688925450) at `5576b6a`, which is ahead of this PR's merge-base. This branch changes no LLM-client or onboarding code; the diff is the per-connection table overlay (`data_source_service`, `tables_schema_section`, one migration) plus tests.

**No fix ported, because none exists in code.** `credit_balance_exhausted` is an account billing state, not a defect — it clears when credits are added to the OpenAI org, and nothing in a PR can make it pass. I have not spent the re-run on it either: a re-run confirms a flake, and a deterministic 429 on an exhausted balance is not one. Happy to re-run once credits are topped up.

For the record on what *was* mine: `unit-tests` had 11 failures on the previous head, of which 2 (`tests/unit/test_refresh_overlay_prefetch.py`) were introduced by this PR and are fixed in `b54a480`. The other 9 (`test_sklearn_sandbox`, `test_artifact_relationship_loading`, `test_read_query_observation_code`) reproduce identically at merge-base `cfe2ea1c` — I ran them there in a worktree to confirm. `e2e-tests (sqlite)`'s 2 failures (`test_rbac_training_mode`, `test_diagnosis_explorer`) are likewise red on `main`.

`unit-tests`, `e2e-tests (sqlite)` and `e2e-tests (postgres)` are still running on this head; I'll follow up if any of them is red for a reason that *is* this PR's.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yochze** [comment]
Re-reviewed commit `b54a48030f12563bf211b7ec7fe7198a8c43462e`. The original reproduction cases now pass, including the prefetch tests. Three remaining findings warrant changes:

1. **[P1] Revocation remains bypassable for migrated overlays.** The migration leaves old user-discovered overlays with `connection_id=NULL`. When connection A becomes denied but B remains overlay-authorized, the predicate admits A's legacy rows through the unconditional NULL allowance. Reproduced through the real selector service: A had a legacy unlinked overlay, A's credentials were deactivated, B retained valid credentials (with no overlay of its own), and the selector still returned `PrivateA`. The new false predicate handles only the case where *no* connection is overlay-authorized. Unknown provenance must not count as authorization on another connection; rediscover under a proven identity or fail closed. The analogous NULL allowance in `SchemaContextBuilder` needs the same ownership rule. [data_source_service.py:3525–3533](https://github.com/bagofwords1/bagofwords/blob/b54a48030f12563bf211b7ec7fe7198a8c43462e/backend/app/services/data_source_service.py#L3525-L3533)

2. **[P2] Switching to service-account mode still serves the personal overlay.** The full-schema merge reads overlays from every connection and filters only by visible table IDs. Even when the classifier correctly puts a connection in the open bucket, its personal overlay is retained and suppresses the canonical schema through `covered`. Reproduced: warm an owner's personal overlay for `Secrets` with only `id`, switch the query identity to `service_account`, verify the connection is classified open, then read the full schema. It still returns only `id`, omitting the service account's `salary` column. Restrict overlay results to overlay-authorized connections so open connections use their canonical schema. [data_source_service.py:3160–3186](https://github.com/bagofwords1/bagofwords/blob/b54a48030f12563bf211b7ec7fe7198a8c43462e/backend/app/services/data_source_service.py#L3160-L3186)

3. **[P2] The roster now merges live and cached client identities.** `_connection_key` uses only the connection ID, collapsing `warehouse` and `warehouse::fast`, which intentionally share a physical connection ID but use different clients. Reproduced through `render_combined`: with a cached table first, a live table on the same connection, another warehouse, and a one-table sample, the roster labels both tables `warehouse::fast`/DuckDB and assigns that alias to `live_orders`; the live `warehouse` identity is absent from the complete render. Use `(connection_id, connection_name)`, matching `_group_tables_by_connection`, whose existing documentation explains why these identities must stay distinct. Preserve the UUID separately for discovery. [tables_schema_section.py:987–990](https://github.com/bagofwords1/bagofwords/blob/b54a48030f12563bf211b7ec7fe7198a8c43462e/backend/app/ai/context/sections/tables_schema_section.py#L987-L990)

Validation: **89 checks passed** across the original probes, updated review/roster/prefetch/migration tests, and adjacent schema, overlay, identity, instruction, and channel tests. **Five PostgreSQL migration checks were skipped** because no test PostgreSQL URL was supplied. **Three new edge-case probes failed** as described above. Service probes used isolated SQLite databases with the external catalog client replaced at `resolve_client_class`; the roster probe exercised rendering. No live-provider or browser QA was run.

**Comment by yochze** [comment]
All three fixed in `2da5154`. Each reproduces at `b54a480` and passes after.

**1 (P1) — revocation bypassable for migrated overlays.** Correct, and my first fix only closed the all-denied case. The underlying rule is that a NULL `connection_id` is *unknown provenance, not permission*, so being authorized on B says nothing about a row that may be A's:

- no connection authorized → nothing
- **any** connection denied → only rows naming an authorized connection
- nothing denied → NULL rows too (no revoked connection for them to have come from, and excluding them would drop a legacy single-connection agent's whole catalog)

Extracted to `_overlay_connection_predicate`, and `SchemaContextBuilder`'s overlay query now calls it instead of carrying its own copy of the allowance — which is how the two drifted apart in the first place.

**2 (P2) — service-account switch still served the personal overlay.** Confirmed. Overlay results are now restricted to overlay-authorized connections under the same rule, so a row that outlives the switch stops suppressing the canonical schema through `covered`. Test asserts `["id"]` while delegated and `["id", "salary"]` once the connection classifies open.

**3 (P2) — roster merged live and `::fast` identities.** Confirmed, and `_group_tables_by_connection`'s docstring already spelled out why they must stay distinct — I should have matched it. Key is now `(connection_id, connection_name)`; the UUID is still emitted separately as `id=` for discovery. At `b54a480` the repro renders `<connections count="2">` with both tables under `warehouse::fast`/duckdb; it now renders 3 with distinct aliases.

**Validation:** 122 passed across the affected suites. The migration suite runs on SQLite **and PostgreSQL 16** — 10/10 with a local PG. Its PostgreSQL params skip when `BOW_TEST_PG_URL` is unset (your 5 skips), and now also skip rather than error when it is set but unreachable.

Separately, on the two red checks: both are the OpenAI account being out of credits, detailed in [the previous comment](https://github.com/bagofwords1/bagofwords/pull/1111#issuecomment-5646744973).

---
_Generated by [Claude Code](https://claude.ai/code)_
